src/oni2-keyboard-layout/stubs/keyboard-layout-linux.c

```diff
@@ -0,0 +1,222 @@
+      bestMatchStyle = thisStyle;
+      break;
+    }
+  }
+
+  XFree(styles);
+  if (bestMatchStyle == 0) {
+    RAISE_NULL;
+  } 
+  
+  Window window;
+  int revertTo;
+  XGetInputFocus(xDisplay, &window, &revertTo);
+
+  if (window != BadRequest) {
+    xInputContext = XCreateIC(
+      xInputMethod,
+      XNInputStyle,
+      bestMatchStyle,
+      XNClientWindow,
+      window,
+      XNFocusWindow,
+      window,
+      NULL
+    );
+  }
+
+  CAMLreturn(Val_unit);
+}
+
+
+void oni2_priv_GetCurrentKeyboardLayout(char *layout) {
+  XkbRF_VarDefsRec vdr;
+  char *tmp;
+  if (XkbRF_GetNamesProp(xDisplay, &tmp, &vdr) && vdr.layout) {
+    XkbStateRec xkbState;
+    XkbGetState(xDisplay, XkbUseCoreKbd, &xkbState);
+
+    if (vdr.variant) {
+      sprintf(layout, "%s,%s[%d]", vdr.layout, vdr.variant, xkbState.group);
+    } else {
+      sprintf(layout, "%s[%d]", vdr.layout, xkbState.group);
+    }
+  } else {
+    layout[0] = '\0';
+  }
+}
+
+CAMLprim value oni2_KeyboardLayoutGetCurrentLayout() {
+  CAMLparam0();
+  CAMLlocal1(vLayout);
+
+  char layout[256];
+  oni2_priv_GetCurrentKeyboardLayout(layout);
+
+  vLayout = caml_copy_string(layout);
+
+  CAMLreturn(vLayout);
+}
+
+CAMLprim value oni2_KeyboardLayoutGetCurrentLanguage() {
+  CAMLparam0();
+  CAMLlocal1(vLanguage);
+
+  char language[256];
+  oni2_priv_GetCurrentKeyboardLayout(language);
+
+  vLanguage = caml_copy_string(language);
+
+  CAMLreturn(vLanguage);
+}
+
+void characterForNativeCode(
+  XIC xInputContext,
+  XKeyEvent *xKeyEvent, 
+  uint xkbKeycode, 
+  uint state,
+  char *buffer
+) {
+  xKeyEvent->keycode = xkbKeycode;
+  xKeyEvent->state = state;
+
+  if (xInputContext) {
+    int count = Xutf8LookupString(xInputContext, xKeyEvent, buffer, 2, NULL, NULL);
+    if (count > 0 && !iswcntrl(buffer[0])) {
+      buffer[count] = '\0';
+    } else {
+      buffer[0] = '\0';
+    }
+  } else {
+    char *cBuffer = (char *)buffer;
+    int count = XLookupString(xKeyEvent, cBuffer, 2, NULL, NULL);
+    if (count > 0 && !iscntrl(cBuffer[0])) {
+      cBuffer[count] = '\0';
+    } else {
+      cBuffer[0] = '\0';
+    }
+  }
+}
+
+CAMLprim value oni2_KeyboardLayoutPopulateCurrentKeymap(value keymap, value Hashtbl_replace) {
+  CAMLparam2(keymap, Hashtbl_replace);
+  CAMLlocal2(keymapEntry, vSDLScancode);
+
+  // Allocate UTF-8 Buffers to be populated
+  char unmodified[4]; 
+  char withShift[4]; 
+  char withAlt[1] = {'\0'};
+
+  XMappingEvent xEventMap = {MappingNotify, 0, false, xDisplay, 0, MappingKeyboard, 0, 0};
+  XRefreshKeyboardMapping(&xEventMap);
+
+  XkbStateRec xkbState;
+  XkbGetState(xDisplay, XkbUseCoreKbd, &xkbState);
+  uint keyboardBaseState = 0x0000;
+  if (xkbState.group == 1) {
+    keyboardBaseState = 0x2000;
+  } else if (xkbState.group == 2) {
+    keyboardBaseState = 0x4000;
+  }
+
+  XEvent xEvent;
+  memset(&xEvent, 0, sizeof(XEvent));
+  XKeyEvent *xKeyEvent = &xEvent.xkey;
+  xKeyEvent->display = xDisplay;
+  xKeyEvent->type = KeyPress;
```

Comment:
Does this `xKeyEvent` need to be allocated?

Reply:
Hmm this is just an alias to the `xEvent` member -- it's not allocating anything on the heap or anything, unless you mean allocating a pointer on the stack. I might be misunderstanding what you mean though